File: Task/Minimal-steps-down-to-1/Go/minimal-steps-down-to-1.go
package main

import (
    "fmt"
    "strings"
)

const limit = 50000

var (
    divs, subs []int
    mins       [][]string
)

// Assumes the numbers are presented in order up to 'limit'.
func minsteps(n int) {
    if n == 1 {
        mins[1] = []string{}
        return
    }
    min := limit
    var p, q int
    var op byte
    for _, div := range divs {
        if n%div == 0 {
            d := n / div
            steps := len(mins[d]) + 1
            if steps < min {
                min = steps
                p, q, op = d, div, '/'
            }
        }
    }
    for _, sub := range subs {
        if d := n - sub; d >= 1 {
            steps := len(mins[d]) + 1
            if steps < min {
                min = steps
                p, q, op = d, sub, '-'
            }
        }
    }
    mins[n] = append(mins[n], fmt.Sprintf("%c%d -> %d", op, q, p))
    mins[n] = append(mins[n], mins[p]...)
}

func main() {
    for r := 0; r < 2; r++ {
        divs = []int{2, 3}
        if r == 0 {
            subs = []int{1}
        } else {
            subs = []int{2}
        }
        mins = make([][]string, limit+1)
        fmt.Printf("With: Divisors: %v, Subtractors: %v =>\n", divs, subs)
        fmt.Println("  Minimum number of steps to diminish the following numbers down to 1 is:")
        for i := 1; i <= limit; i++ {
            minsteps(i)
            if i <= 10 {
                steps := len(mins[i])
                plural := "s"
                if steps == 1 {
                    plural = " "
                }
                fmt.Printf("    %2d: %d step%s: %s\n", i, steps, plural, strings.Join(mins[i], ", "))
            }
        }
        for _, lim := range []int{2000, 20000, 50000} {
            max := 0
            for _, min := range mins[0 : lim+1] {
                m := len(min)
                if m > max {
                    max = m
                }
            }
            var maxs []int
            for i, min := range mins[0 : lim+1] {
                if len(min) == max {
                    maxs = append(maxs, i)
                }
            }
            nums := len(maxs)
            verb, verb2, plural := "are", "have", "s"
            if nums == 1 {
                verb, verb2, plural = "is", "has", ""
            }
            fmt.Printf("  There %s %d number%s in the range 1-%d ", verb, nums, plural, lim)
            fmt.Printf("that %s maximum 'minimal steps' of %d:\n", verb2, max)
            fmt.Println("   ", maxs)
        }
        fmt.Println()
    }
}

File: Task/Minimal-steps-down-to-1/C++/minimal-steps-down-to-1.cpp
#include <cstdint>
#include <iomanip>
#include <iostream>
#include <string>
#include <vector>

const int32_t limit = 50'000;

std::vector<int32_t> divisors;
std::vector<int32_t> subtractors;
std::vector<std::vector<std::string>> minimums;

template <typename T>
void print_vector(const std::vector<T>& list) {
	for ( uint64_t i = 0; i < list.size(); ++i ) {
		std::cout << list[i];
		if ( i < list.size() - 1 ) {
			std::cout << ", ";
		}
	}
}

// Assumes that numbers are presented in ascending order up to 'limit'.
void minimum_steps(int32_t n) {
    if ( n == 1 ) {
        return;
    }

    int32_t minimum = limit;
	int32_t p = 0;
	int32_t q = 0;
	std::string operator_symbol = "";
	for ( int32_t divisor : divisors ) {
		if ( n % divisor == 0 ) {
		    int32_t d = n / divisor;
			int32_t steps = minimums[d].size() + 1;
			if ( steps < minimum ) {
				minimum = steps;
				p = d;
				q = divisor;
				operator_symbol = "/";
			}
		}
	}

	for ( int32_t subtractor : subtractors ) {
		int32_t d = n - subtractor;
		if ( d >= 1 ) {
			int32_t steps = minimums[d].size() + 1;
			if ( steps < minimum ) {
				minimum = steps;
				p = d;
				q = subtractor;
				operator_symbol = "-";
			}
		}
	}

	minimums[n].emplace_back(operator_symbol + std::to_string(q) + " -> " + std::to_string(p));
	minimums[n].insert(minimums[n].end(), minimums[p].begin(), minimums[p].end());
}

int main() {
	for ( int32_t item : { 0, 1 } ) {
	    divisors = { 2, 3 };
	    subtractors = { item + 1 };
	    minimums = std::vector(limit + 1, std::vector<std::string>(0));
	    std::cout << "With: Divisors: { "; print_vector<int32_t>(divisors);
	    std::cout << " }, Subtractors: { "; print_vector<int32_t>(subtractors); std::cout << " } =>" << std::endl;
	    std::cout << "  Minimum number of steps to diminish the following numbers down to 1 is:" << std::endl;
	    for ( int32_t i = 1; i < limit; ++i ) {
	        minimum_steps(i);
	        if ( i <= 10 ) {
	            int32_t steps = minimums[i].size();
	            const std::string plural = ( steps == 1 ) ? " : " : "s: ";
	            std::cout << "    " << std::setw(2) << i << ": " << steps << " step" + plural;
	            print_vector<std::string>(minimums[i]); std::cout << std::endl;
	        }
	    }

	    for ( int32_t lim : { 2'000, 20'000, 50'000 } ) {
		   uint64_t max = 0;
		   for ( int32_t j = 1; j <= lim; ++j ) {
			   uint64_t m = minimums[j].size();
			   if ( m > max ) {
				   max = m;
			   }
		   }
		   std::vector<int32_t> maxs;
		   for ( int32_t j = 1; j <= lim; ++j ) {
			   if ( minimums[j].size() == max ) {
				   maxs.emplace_back(j);
			   }
		   }

		   int32_t size = maxs.size();
		   std::string verb1  = ( size == 1 ) ? "is" : "are";
		   std::string verb2  = ( size == 1 ) ? "has" : "have";
		   std::string plural = ( size == 1 ) ? "" : "s";
		   std::cout << "  There " << verb1 << " " << size << " number" << plural << " in the range 1 - " << lim
				     << " that " << verb2 << " maximum 'minimal steps' of " << max << ":" << std::endl;
		   std::cout << "  [ "; print_vector<int32_t>(maxs); std::cout << " ]" << std::endl;
	   }
	   std::cout << std::endl;
	}
}
